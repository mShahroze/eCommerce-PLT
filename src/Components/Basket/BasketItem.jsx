import React, { useState } from "react";
import "../Styles/BasketItem.css";
import { connect } from "react-redux";
import {
  removeFromBasket,
  adjustItemQuantity,
} from "../../Redux/Shopping/shopping-actions";

const BasketItem = ({ item, removeFromBasket, adjustQuantity }) => {
  const [input, setInput] = useState(item.quantity);
  const onChangeHandler = (e) => {
    setInput(e.target.value);
    adjustQuantity(item.id, e.target.value);
  };

  return (
    <div data-testid="basketItem" className="basketItem">
      <img className="basketItemImg" src={item.img} alt={item.name} />
      <div className="basketItemDetails">
        <p className="detailsName">{item.name}</p>
        <p className="detailsColour">{item.colour}</p>
        <p className="detailsPrice">Â£ {item.price}</p>
      </div>
      <div className="basketItemActions">
        <div className="basketItemQuantity">
          <label htmlFor="qty">Qty</label>
          <input
            data-testid="input"
            min="1"
            type="number"
            id="qty"
            name="qty"
            value={input}
            onChange={onChangeHandler}
          />
        </div>
        <button
          data-testid="removeItem"
          onClick={() => removeFromBasket(item.id)}
          className="actionsDeleteItemButton"
        >
          <img
            src="https://image.flaticon.com/icons/svg/709/709519.svg"
            alt=""
          />
        </button>
      </div>
    </div>
  );
};

const mapDispatchToProps = (dispatch) => {
  return {
    adjustQuantity: (id, quantity) =>
      dispatch(adjustItemQuantity(id, quantity)),
    removeFromBasket: (id) => dispatch(removeFromBasket(id)),
  };
};

export default connect(null, mapDispatchToProps)(BasketItem);
