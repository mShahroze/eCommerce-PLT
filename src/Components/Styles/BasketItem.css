.basketItem {
  margin-bottom: 1rem;
  display: flex;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.26);
  border-radius: 10px;
  height: 251px;
}

.basketItemImg {
  width: 350px;
  object-fit: contain;
  border-radius: 10px;
}

.basketItemDetails {
  padding: 1rem;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  color: var(--secondary-color);
}

.detailsName {
  font-size: 1rem;
  font-weight: bold;
}

.detailsColour {
  font-size: 0.8rem;
}

.detailsPrice {
  font-size: 1rem;
  font-weight: bold;
}

.basketItemActions {
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  align-items: center;
  padding: 0.6rem;
}

.basketItemQuantity {
  display: flex;
  align-items: center;
}

.basketItemQuantity input {
  padding: 10px;
  margin-left: 0.4rem;
  width: 60px;
  border-radius: 10px;
  border: 1px solid var(--secondary-color);
  outline: none;
}

.actionsDeleteItemButton {
  width: 50px;
  height: 50px;
  border: 1px solid var(--secondary-color);
  border-radius: 50%;
  cursor: pointer;
  transition: all ease-in-out 0.7s;
  outline: none;
}

.actionsDeleteItemButton:hover {
  transform: scale(1.2) rotate(360deg);
  background-color: rgb(209, 15, 15);
}

.actionsDeleteItemButton img {
  width: 25px;
  height: 25px;
}
