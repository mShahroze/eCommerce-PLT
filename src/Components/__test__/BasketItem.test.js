import React from "react";
import BasketItem from "../Basket/BasketItem";
import { fireEvent, render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {
  shop: {
    products: [],
    basket: [
      {
        id: 1,
        colour: "Black",
        name: "Black Sheet Strappy Textured Glitter Bodycon Dress",
        price: 10,
        img: "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
        quantity: 1,
      },
    ],
    currentItem: null,
    loading: true,
  },
};
const store = mockStore(initialState);

describe("BasketItem", () => {
  test("renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <BasketItem item={initialState.shop.basket[0]} />
        </BrowserRouter>
      </Provider>
    );

    expect(container);
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <BasketItem item={initialState.shop.basket[0]} />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test("change value of quantity input work correctly", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <BrowserRouter>
          <BasketItem item={initialState.shop.basket[0]} />
        </BrowserRouter>
      </Provider>
    );
    const inputEl = getByTestId("input");
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation((init) => [init, setState]);

    expect(inputEl.value).toBe("1");

    fireEvent.change(inputEl, {
      target: {
        value: 5,
      },
    });
    expect(inputEl.value).toBe("5");
  });
});
