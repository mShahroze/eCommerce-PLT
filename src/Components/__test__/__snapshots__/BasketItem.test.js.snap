// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasketItem matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="basketItem"
    data-testid="basketItem"
  >
    <img
      alt="Black Sheet Strappy Textured Glitter Bodycon Dress"
      class="basketItemImg"
      src="http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024"
    />
    <div
      class="basketItemDetails"
    >
      <p
        class="detailsName"
      >
        Black Sheet Strappy Textured Glitter Bodycon Dress
      </p>
      <p
        class="detailsColour"
      >
        Black
      </p>
      <p
        class="detailsPrice"
      >
        Â£ 10
      </p>
    </div>
    <div
      class="basketItemActions"
    >
      <div
        class="basketItemQuantity"
      >
        <label
          for="qty"
        >
          Qty
        </label>
        <input
          data-testid="input"
          id="qty"
          min="1"
          name="qty"
          type="number"
          value="1"
        />
      </div>
      <button
        class="actionsDeleteItemButton"
        data-testid="removeItem"
      >
        <img
          alt=""
          src="https://image.flaticon.com/icons/svg/709/709519.svg"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
